verifier/verify.go

```diff
@@ -125,86 +129,179 @@ func (v *verifier) Stop() {
-				response.Verified = false
-				response.Message = fmt.Sprintf("failed to verify ephemeral container: %s: %s", image.Name, err.Error())
-				break
-			}
-			response.Results = append(response.Results, *result)
+	if err := response.BuildAttestationList(requestData.Attestations); err != nil {
+		return nil, errors.Wrapf(err, "failed to create attestation list")
+	}
+
+	response.ResponseData.Attestations, err = v.verifyAttestations(ctx, response.GetImageList())
+	if err != nil {
+		return response.VerificationFailed(fmt.Sprintf("failed to verify attestatations: %v", err.Error()))
+	}
+
+	return response.VerificationSucceeded("")
+}
+
+func (v *verifier) verifyAttestations(ctx context.Context, imageList map[string]AttestationList) ([]Attestation, error) {
+	attestations := make([]Attestation, 0)
+	for image, list := range imageList {
+		imageAttestations, err := v.verifyAttestation(ctx, image, list)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to verify attestations")
 		}
-		v.logger.Infof("verified %d ephemeralContainers", images.EphemeralContainers)
+		attestations = append(attestations, imageAttestations...)
+	}
+	return attestations, nil
+}
+
+func (v *verifier) verifyAttestation(ctx context.Context, image string, attestationList AttestationList) ([]Attestation, error) {
+	if len(attestationList) == 0 {
+		return []Attestation{}, nil
+	}
+
+	attestations := make([]Attestation, len(attestationList))
+	remoteOpts, err := v.getRemoteOpts(ctx, image)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get gcr remote opts")
+	}
+
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse image reference: %s", image)
+	}
+
+	refDesc, err := gcrremote.Head(ref, remoteOpts...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get gcr remote head")
+	}
+
+	referrers, err := gcrremote.Referrers(ref.Context().Digest(refDesc.Digest.String()), remoteOpts...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get gcr remote referrers")
+	}
+
+	referrersDescs, err := referrers.IndexManifest()
+	if err != nil {
+		return nil, err
+	}
+
+	for i, referrer := range referrersDescs.Manifests {
+		if !attestationList[referrer.ArtifactType] {
+			continue
+		}
+
+		referrerRef := v.getReference(referrer, ref)
+
+		_, err := v.verifyReferences(ctx, referrerRef)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get referrer of artifact type %s", referrer.ArtifactType)
+		}
+
+		payload, err := v.extractPayload(ctx, ref, referrer, remoteOpts...)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to extract payload")
+		}
+
+		attestations[i] = Attestation{
+			Type:    referrer.ArtifactType,
+			Image:   referrerRef,
+			Payload: payload,
+		}
+	}
+	return attestations, nil
+}
+
+func (v *verifier) extractPayload(ctx context.Context, repoRef name.Reference, desc v1.Descriptor, options ...gcrremote.Option) (map[string]interface{}, error) {
```

Comment:
If this method is also doing a fetch, we can perhaps rename it to indicate that? e.g. `fetchPayload` or `fetchAndExtractPayload`?